crates/node-bindings/src/resolver.rs

```diff
@@ -395,6 +408,43 @@ impl Resolver {
     }
   }
 
+  #[napi]
+  pub fn resolve(&self, options: ResolveOptions, env: Env) -> Result<ResolveResult> {
+    let (res, side_effects, module_type) = self.resolve_internal(options)?;
+    self.resolve_result_to_js(env, res, side_effects, module_type)
+  }
+
+  #[cfg(target_arch = "wasm32")]
+  #[napi]
+  pub fn resolve_async(&'static self) -> Result<JsObject> {
+    panic!("resolveAsync() is not supported in Wasm builds")
+  }
+
+  #[cfg(not(target_arch = "wasm32"))]
+  #[napi]
+  pub fn resolve_async(&'static self, options: ResolveOptions, env: Env) -> Result<JsObject> {
+    let (deferred, promise) = env.create_deferred()?;
+    let resolver = &self.resolver;
+
+    if matches!(resolver.cache.fs, EitherFs::A(..)) || resolver.module_dir_resolver.is_some() {
```

Comment:
I tried to understand how the EitherFS thing works, but I don't think I fully have my head around it.

It's an enum with two generic params, and the value of the enum is either of the generics.. and the generic type of the Fs used for the resolver is EitherFS<JsFileSystem, OsFileSystem> - does that just mean that the fs has to be one of those two? I'm sure there's a good reason (that I'd love to know more about), but why can't the type just be `JsFileSystem | OsFilesSystem`?

It's a bit off topic, but I'd love to learn more.

Reply:
In rust there are no union types like that. EitherFs is a generic enum that chooses between (any) two options and proxies all of the FileSystem methods to the underlying instance.